### ISSUE TITLE
Can't be stored in react's setState.

### ISSUE THREAD

**Issue by safareli** [NiGhTTraX/strong-mock#322]
[This is minimal reproduction](https://codesandbox.io/s/hungry-bas-zk5ykn?file=/src/App.js:0-374):

```jsx
import { useState } from "react";
import * as SM from "strong-mock";
import "./styles.css";

export default function App() {
  const [_state, setState] = useState(null);
  return (
    <div className="App">
      <button
        onClick={() => {
          const foo = SM.mock();
          setState(foo);
        }}
      >
        setState
      </button>
    </div>
  );
}

```

![Screenshot 2023-02-23 at 15 58 00](https://user-images.githubusercontent.com/1932383/220899341-12b41702-d4ea-445b-8601-6686e5f2b688.png)

**Comment by NiGhTTraX** [comment]
Hey, @safareli, thanks for raising this! Unfortunately, I don't see a way of addressing this. Like you said, [React](https://github.com/facebook/react/blob/c04b18070145b82111e1162729f4776f4d2c6112/packages/react-reconciler/src/ReactFiberHooks.js#L1029) will check if the value is a function, and all strong-mocks are, because the [Proxy target](https://github.com/NiGhTTraX/strong-mock/blob/d70b6cc0e1cdd8cf5f071646dc10cc64272a8272/src/proxy.ts#L55) is set to a function. This is REQUIRED, otherwise strong-mock can't intercept function calls at all.

#321 was easily solvable because `Promise.resolve` looks for a `.then` property, so I added that to [the list](https://github.com/NiGhTTraX/strong-mock/blob/04afc77fea4662580c2651e13ec1783d067fe3b1/src/expectation/repository/flexible-repository.ts#L107-L124) of properties strong-mock returns a default value for.

As much as I try to make strong-mock not interfere with production code, this might be a case where you have to adapt it a bit, e.g. explicitly using the callback version of `setState`, or putting the mocked function inside an object.

**Comment by safareli** [comment]
Thanks, that makes sense.

Could there be another version of `mock`  function that is not using proxy of a function under the hood? maybe just an object instead. Then if user is fine with not intercepting function calls they would use this variation?

mock could be accepting optional value that would be used for making proxy for example.

**Comment by NiGhTTraX** [comment]
If you don't intercept any function calls, then you might as well just fake the object directly, rather than mocking it.

```ts
const foo = { bar: 42 };
```

is simpler than

```ts
const foo = mock<{ bar: number }>();
when(() => foo.bar).thenReturn(42);
```

**Comment by golergka** [comment]
> If you don't intercept any function calls

Unfortunately, we indeed need to intercept calls on mocked object in this case.

**Comment by NiGhTTraX** [comment]
Closed by #323.
